server/viame_server/serializers/viame.py

```diff
@@ -92,50 +147,24 @@ def load_csv_as_tracks(file):
     tracks = {}
 
     for row in reader:
-        features, attributes, track_attributes, confidence_pairs = _parse_row(row)
-        trackid = int(row[0])
-        frame = int(row[2])
-        bounds = [
-            float(row[3]),
-            float(row[5]),
-            float(row[4]),
-            float(row[6]),
-        ]
-        fishLength = float(row[8])
-
-        if trackid not in tracks:
-            # track is defined as follows...
-            tracks[trackid] = {
-                # First frame with a feature
-                "begin": frame,
-                # Last frame with a feature
-                "end": frame,
-                # Unique among tracks in the video
-                "key": trackid,
-                # Array<{{ frame: number, foo: bar, ...}}>
-                "features": [],
-                # Key is an attribute name, val is a float confidence value
-                "confidencePairs": {},
-                # Key is string, value is freeform
-                "attributes": {},
-            }
-        track = tracks[trackid]
-        track["begin"] = min(frame, track["begin"])
-        features["frame"] = frame
-        features["bounds"] = bounds
-        if fishLength > 0:
-            features["fishLength"] = fishLength
-
-        if attributes:
-            features["attributes"] = attributes
-        track["features"].append(features)
+        (
+            feature,
+            attributes,
+            track_attributes,
+            confidence_pairs,
+        ) = _parse_row_for_tracks(row)
+        trackId, frame, _, _ = row_info(row)
+
+        if trackId not in tracks:
+            tracks[trackId] = Track(frame, frame, trackId)
+
+        track = tracks[trackId]
+        track.begin = min(frame, track.begin)
+        track.end = max(track.end, frame)
+        track.features.append(feature)
+        track.confidencePairs = confidence_pairs
+
         for (key, val) in track_attributes:
-            track["attributes"][key] = val
-        if frame > track["end"]:
-            track["end"] = frame
-            # final confidence pair should be taken as the
-            # pair that applied to the whole track
-            for (key, val) in confidence_pairs:
-                track["confidencePairs"][key] = val
-
-    return tracks
+            track.attributes[key] = val
+
+    return {trackId: asdict(track) for trackId, track in tracks.items()}
```

Comment:
This results in a different data object than the code it's replacing, and it's different than the client expects:

```
      {
        "attributes": {},
        "bounds": [
          654,
          743,
          173,
          266
        ],
        "fishLength": null,
        "frame": 2,
        "head": null,
        "tail": null
      },
```

Before, where a particular feature type is missing, the key will simply be omitted.  `feature.head` is never null, it's either a number or undefined.

Reply:
Also, `asdict` makes serialization slower as the entire dict has to be constructed from scratch: https://stackoverflow.com/questions/52229521/why-is-dataclasses-asdictobj-10x-slower-than-obj-dict

I'm actually not sure what value we're getting out of dataclasses, now that I think about it.  I think the idea is that they can be used when you need to make your structs hashable or comparable, but we aren't using any of that....

Reply:
The main benefit I get from dataclasses are:
- Description and type definition of our JSON
- Being able to work with our data in a more object oriented and clean way.

FWIW having `null`/`None` things be omitted is an easy thing to do, and I could try to replace the `asdict` call with calls to `__dict__`, but that may not work as I believe there are nested dataclasses in this PR. 

If we have serious performance concerns but still want to keep the benefit of typed data (which I think we should), we could switch to using `TypedDict` with `mypy`. Generally I think `mypy` would be a good addition anyway. My main point is that I'm not a fan of passing around these dictionaries willy nilly without any concrete definition of them somewhere. IMO, `TypedDict` would fill that roll, and wouldn't have a runtime performance hit, as they're purely for static type checking and are just a `Dict` at runtime. I think it would look a bit less clean than what I've proposed, but I'm also shrugging here.

Reply:
@subdavis ef22b41 addresses your concerns around using `asdict`, lmk what you think.